File: WebLabelPrint_VB/Models/ServerPrinterInfo.vb

Imports System.Collections.Generic
Imports System.Linq
Imports System.Management
Imports System.Web
Imports System.Web.Script.Serialization

Namespace WebLabelPrint.Models
   ''' <summary>
   ''' Contains information about printers installed on the server.
   ''' </summary>
   Public Class ServerPrinterInfo
      Public Property PrinterName() As String
         Get
            Return m_PrinterName
         End Get
         Set(value As String)
            m_PrinterName = value
         End Set
      End Property
      Private m_PrinterName As String
      Public Property DriverName() As String
         Get
            Return m_DriverName
         End Get
         Set(value As String)
            m_DriverName = value
         End Set
      End Property
      Private m_DriverName As String
      Public Property PortName() As String
         Get
            Return m_PortName
         End Get
         Set(value As String)
            m_PortName = value
         End Set
      End Property
      Private m_PortName As String
      Public Property Location() As String
         Get
            Return m_Location
         End Get
         Set(value As String)
            m_Location = value
         End Set
      End Property
      Private m_Location As String
      Public Property IsDefault() As Boolean
         Get
            Return m_IsDefault
         End Get
         Set(value As Boolean)
            m_IsDefault = value
         End Set
      End Property
      Private m_IsDefault As Boolean

      ''' <summary>
      ''' Exports a JSON version of the server's printer information.
      ''' </summary>
      Public Function ExportToJson() As String
         Return New JavaScriptSerializer().Serialize(Me)
      End Function

      ''' <summary>
      ''' Performs a WMI query to return information about printers installed on the server.
      ''' 
      ''' We recommend that you cache the results in memory as this can be slow particularly when
      ''' querying network-connected printers.
      ''' </summary>
      Shared Function GetServerPrinters() As List(Of ServerPrinterInfo)
         Dim serverPrintersList As New List(Of ServerPrinterInfo)()

         Dim query As String = "SELECT Name,PortName,Location,DriverName,Default FROM Win32_Printer"

         Dim searcher As New ManagementObjectSearcher(query)
         Dim collection As ManagementObjectCollection = searcher.[Get]()
         For Each printer As ManagementObject In collection
            ' Printer Name
            Dim printerName As String = String.Empty
            If (printer.Properties("Name") IsNot Nothing) AndAlso (printer.Properties("Name").Value IsNot Nothing) Then
               printerName = printer.Properties("Name").Value.ToString()
            End If

            ' Driver Name (Model Name)
            Dim printerModel As String = String.Empty
            If (printer.Properties("DriverName") IsNot Nothing) AndAlso (printer.Properties("DriverName").Value IsNot Nothing) Then
               printerModel = printer.Properties("DriverName").Value.ToString()
            End If

            ' Port Name
            Dim printerPort As String = String.Empty
            If (printer.Properties("PortName") IsNot Nothing) AndAlso (printer.Properties("PortName").Value IsNot Nothing) Then
               printerPort = printer.Properties("PortName").Value.ToString()
            End If

            ' Location
            Dim printerLocation As String = String.Empty
            If (printer.Properties("Location") IsNot Nothing) AndAlso (printer.Properties("Location").Value IsNot Nothing) Then
               printerLocation = printer.Properties("Location").Value.ToString()
            End If

            ' IsDefault
            Dim isDefault As Boolean = False
            If (printer.Properties("Default") IsNot Nothing) AndAlso (printer.Properties("Default").Value IsNot Nothing) Then
               isDefault = CBool(printer.Properties("Default").Value)
            End If

            Dim printerInfo As New ServerPrinterInfo()
            printerInfo.PrinterName = printerName
            printerInfo.DriverName = printerModel
            printerInfo.PortName = printerPort
            printerInfo.Location = printerLocation
            printerInfo.IsDefault = isDefault

            serverPrintersList.Add(printerInfo)
         Next

         Return serverPrintersList
      End Function
   End Class
End Namespace

File: WebLabelPrint_CS/Models/ServerPrinterInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Management;
using System.Web;
using System.Web.Script.Serialization;

namespace WebLabelPrint.Models
{
   /// <summary>
   /// Contains information about printers installed on the server.
   /// </summary>
   public class ServerPrinterInfo
   {
      public string PrinterName { get; set; }
      public string DriverName { get; set; }
      public string PortName { get; set; }
      public string Location { get; set; }
      public bool IsDefault { get; set; }

      /// <summary>
      /// Exports a JSON version of the server's printer information.
      /// </summary>
      public string ExportToJson()
      {
         return new JavaScriptSerializer().Serialize(this);
      }

      /// <summary>
      /// Performs a WMI query to return information about printers installed on the server.
      /// 
      /// We recommend that you cache the results in memory as this can be slow particularly when
      /// querying network-connected printers.
      /// </summary>
      public static List<ServerPrinterInfo> GetServerPrinters()
      {
         List<ServerPrinterInfo> serverPrintersList = new List<ServerPrinterInfo>();

         string query = "SELECT Name,PortName,Location,DriverName,Default FROM Win32_Printer";

         ManagementObjectSearcher searcher = new ManagementObjectSearcher(query);
         ManagementObjectCollection collection = searcher.Get();
         foreach (ManagementObject printer in collection)
         {
            // Printer Name
            string printerName = string.Empty;
            if ((printer.Properties["Name"] != null) && (printer.Properties["Name"].Value != null))
               printerName = printer.Properties["Name"].Value.ToString();
               
            // Driver Name (Model Name)
            string printerModel = string.Empty;
            if ((printer.Properties["DriverName"] != null) && (printer.Properties["DriverName"].Value != null))
               printerModel = printer.Properties["DriverName"].Value.ToString();

            // Port Name
            string printerPort = string.Empty;
            if ((printer.Properties["PortName"] != null) && (printer.Properties["PortName"].Value != null))
               printerPort = printer.Properties["PortName"].Value.ToString();

            // Location
            string printerLocation = string.Empty;
            if ((printer.Properties["Location"] != null) && (printer.Properties["Location"].Value != null))
               printerLocation = printer.Properties["Location"].Value.ToString();

            // IsDefault
            bool isDefault = false;
            if ((printer.Properties["Default"] != null) && (printer.Properties["Default"].Value != null))
               isDefault = ((bool)printer.Properties["Default"].Value);

            ServerPrinterInfo printerInfo = new ServerPrinterInfo();
            printerInfo.PrinterName = printerName;
            printerInfo.DriverName = printerModel;
            printerInfo.PortName = printerPort;
            printerInfo.Location = printerLocation;
            printerInfo.IsDefault = isDefault;

            serverPrintersList.Add(printerInfo);
         }

         return serverPrintersList;
      }
   }
}